File: app/src/main/java/no/joharei/flixr/photosets/PhotosetsFragment.java
package no.joharei.flixr.photosets;


import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.v17.leanback.app.VerticalGridFragment;
import android.support.v17.leanback.widget.ArrayObjectAdapter;
import android.support.v17.leanback.widget.VerticalGridPresenter;

import com.f2prateek.dart.Dart;
import com.f2prateek.dart.InjectExtra;

import java.util.List;

import no.joharei.flixr.CardPresenter;
import no.joharei.flixr.Henson;
import no.joharei.flixr.api.models.Photoset;

public class PhotosetsFragment extends VerticalGridFragment implements PhotosetsView {

    private static final int NUM_COLUMNS = 4;
    @InjectExtra
    String userId;
    private ArrayObjectAdapter mAdapter;
    private PhotosetsPresenter photosetsPresenter;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        photosetsPresenter = new PhotosetsPresenter();
        photosetsPresenter.attachView(this);

        VerticalGridPresenter gridPresenter = new VerticalGridPresenter();
        gridPresenter.setNumberOfColumns(NUM_COLUMNS);
        setGridPresenter(gridPresenter);

        mAdapter = new ArrayObjectAdapter(new CardPresenter());
        setAdapter(mAdapter);
        setOnItemViewClickedListener((itemViewHolder, item, rowViewHolder, row) -> {
            if (item instanceof Photoset) {
                Photoset photoset = (Photoset) item;
                Intent intent = Henson.with(getActivity())
                        .gotoPhotosActivity()
                        .photosetId(photoset.getId())
                        .userId(userId)
                        .build();
                getActivity().startActivity(intent);
            }
        });

        loadPhotosets();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        Dart.inject(this, getActivity());
    }

    private void loadPhotosets() {
        photosetsPresenter.fetchPhotosets(userId);
    }

    @Override
    public void showPhotosets(List<Photoset> photosets) {
        mAdapter.addAll(0, photosets);
    }
}

File: app/src/main/java/no/joharei/flixr/photosets/PhotosetsFragment.kt
package no.joharei.flixr.photosets


import android.app.ProgressDialog
import android.content.Context
import android.os.Bundle
import androidx.annotation.Nullable
import androidx.leanback.app.VerticalGridSupportFragment
import androidx.leanback.widget.ArrayObjectAdapter
import androidx.leanback.widget.VerticalGridPresenter
import com.f2prateek.dart.Dart
import com.f2prateek.dart.InjectExtra
import no.joharei.flixr.CardPresenter
import no.joharei.flixr.Henson
import no.joharei.flixr.network.models.Photoset

class PhotosetsFragment : VerticalGridSupportFragment(), PhotosetsView {
    @InjectExtra
    @Nullable
    @JvmField
    internal var userId: String? = null
    private val photosetsAdapter by lazy { ArrayObjectAdapter(CardPresenter(context)) }
    private val photosetsPresenter = PhotosetsPresenter()
    private val progressDialog by lazy { ProgressDialog(context) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        photosetsPresenter.attachView(this)

        val gridPresenter = VerticalGridPresenter()
        gridPresenter.numberOfColumns = 4
        setGridPresenter(gridPresenter)

        adapter = photosetsAdapter
        setOnItemViewClickedListener { _, item, _, _ ->
            if (item is Photoset) {
                val intent = Henson.with(activity)
                        .gotoPhotosActivity()
                        .photosetId(item.id)
                        .photosetTitle(item.title)
                        .userId(userId)
                        .build()
                startActivity(intent)
            }
        }

        loadPhotosets()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        Dart.inject(this, activity)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        photosetsPresenter.stop()
    }

    private fun loadPhotosets() {
        photosetsPresenter.fetchPhotosets(userId)
    }

    override fun showPhotosets(photosets: List<Photoset>) {
        photosetsAdapter.addAll(0, photosets)
    }

    override fun showProgress() {
        progressDialog.setTitle("Please wait")
        progressDialog.setMessage("Loading photosets...")
        progressDialog.show()
    }

    override fun hideProgress() {
        progressDialog.dismiss()
    }

    override fun getContext(): Context = requireContext()

}
